### PR TITLE
auth: Adding unit tests

### PR DESCRIPTION
**Pull request by Rushit**
This covers tests for User and Role related operations. This tests brings code coverage in store.go from 40.2% to 72.3%.

/cc @mitake

### CODE REVIEW

File: auth/store_test.go [hunk 1]
```diff
@@ -230,6 +231,119 @@ func TestRoleAdd(t *testing.T) {
...
+
+	// adds a new role
+	_, roleAddErr = as.RoleAdd(&pb.AuthRoleAddRequest{Name: "role-test"})
+	if roleAddErr != nil {
+		t.Fatal(roleAddErr)
+	}
+	as.UserAdd(&pb.AuthUserAddRequest{Name: "user1", Password: "xyz"})
+	as.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: "user1", Role: "role-test"})
+
+	t.Run("Test Get User", func(t *testing.T) {
```
Comment (mitake) [inline auth/store_test.go]: Thanks for enhancing the test but each test should be separated into individual functions e.g. `TestGetUser()`. Could you separate them?

File: .gitignore [hunk 2]
```diff
@@ -12,3 +12,4 @@
 *.test
 tools/functional-tester/docker/bin
 hack/tls-setup/certs
+.idea
```
Comment (mitake) [inline .gitignore]: What is this?

File: auth/store_test.go (hunk 1 above)
Comment (Rushit) [inline auth/store_test.go]: sure. I can do it.

File: .gitignore (hunk 2 above)
Comment (Rushit) [inline .gitignore]: Intellij Idea IDE specific files. Not sure whether it's right way to do it or not?

Comment (mitake) [inline .gitignore]: I'm not sure adding it to .gitignore is a good idea or not. But if you want to do this, could you separate a commit and create another PR for the change? It is not related to auth unit tests.

Comment (Rushit) [inline .gitignore]: yes. I'll put it into another PR.

Comment (Rushit) [inline .gitignore]: https://github.com/coreos/etcd/pull/7230/files

File: auth/store_test.go (hunk 1 above)
Comment (Rushit) [inline auth/store_test.go]: done.

File: auth/store_test.go [hunk 3]
```diff
@@ -15,13 +15,14 @@
...
-	"os"
-	"testing"
-
+	"github.com/coreos/etcd/auth/authpb"
 	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
 	"github.com/coreos/etcd/mvcc/backend"
+	"github.com/stretchr/testify/assert"
 	"golang.org/x/crypto/bcrypt"
 	"golang.org/x/net/context"
+	"os"
```
Comment (xiang90) [inline auth/store_test.go]: please change back to the previous import blocks.

```
standard

internal import

external import
```

File: auth/store_test.go [hunk 4]
```diff
@@ -230,6 +231,126 @@ func TestRoleAdd(t *testing.T) {
 	}
 }
 
+func setupAuthStore(t *testing.T) (AuthStore, func(t *testing.T)) {
```
Comment (xiang90) [inline auth/store_test.go]: if we want to have setup, we should user it everywhere in this pkg. we need to make other tests consistent. probably you can submit a pr to make this happen first.

File: auth/store_test.go [hunk 5]
```diff
@@ -230,6 +231,126 @@ func TestRoleAdd(t *testing.T) {
...
+}
+
+func TestGetUser(t *testing.T) {
+	as, tearDown := setupAuthStore(t)
+	defer tearDown(t)
+	resp, userGetErr := as.UserGet(&pb.AuthUserGetRequest{Name: "user1"})
+	if userGetErr != nil {
+		t.Fatal(userGetErr)
+	}
+	assert.Equal(t, []string{"role-test"}, resp.Roles, "It should not have any roles")
```
Comment (xiang90) [inline auth/store_test.go]: please avoid using assert. use simple if else instead. thanks.

File: auth/store_test.go [hunk 6]
```diff
@@ -230,6 +231,126 @@ func TestRoleAdd(t *testing.T) {
...
+	_, delErr := as.RoleDelete(&pb.AuthRoleDeleteRequest{Role: "role-test"})
+	if delErr != nil {
+		t.Fatal(delErr)
+	}
+	listResp, listError := as.RoleList(&pb.AuthRoleListRequest{})
+	if listError != nil {
+		t.Fatal(listError)
+	}
+	assert.Equal(t, []string{"root"}, listResp.Roles, "It should not have any roles")
+
```
Comment (xiang90) [inline auth/store_test.go]: remove empty line here.

File: auth/store_test.go [hunk 7]
```diff
@@ -230,6 +231,126 @@ func TestRoleAdd(t *testing.T) {
...
+		Key:      "Keys",
+		RangeEnd: "RangeEnd",
+	})
+	if revokeErr != nil {
+		t.Error(revokeErr)
+	}
+
+	resp, getErr = as.RoleGet(&pb.AuthRoleGetRequest{Role: "role-test-1"})
+	assert.Equal(t, 0, len(resp.Perm), "It should not have any permissions.")
+
```
Comment (xiang90) [inline auth/store_test.go]: remove empty line.

**Comment by gyuho** [comment]
Also we don't want to have separate merge commits in the PR. Could you also rebase your branch to current etcd master? Thanks.

File: auth/store_test.go (hunk 5 above)
Comment (Rushit) [inline auth/store_test.go]: Sure, I can change it if..else. I have seen it used in ./store/state_test.go. Just curious for my knowledge, why it is not recommended practice?

File: auth/store_test.go (hunk 4 above)
Comment (Rushit) [inline auth/store_test.go]: i'll submit refactored PR first.

File: auth/store_test.go (hunk 3 above)
Comment (Rushit) [inline auth/store_test.go]: got it.

File: auth/store_test.go (hunk 4 above)
Comment (Rushit) [inline auth/store_test.go]: https://github.com/coreos/etcd/pull/7234/files

**Comment by Rushit** [comment]
Hoping to merge https://github.com/coreos/etcd/pull/7234 first. So I can use that branch to add new tests.

File: auth/store_test.go [hunk 8]
```diff
@@ -18,8 +18,11 @@ import (
 	"os"
 	"testing"
 
+	"github.com/coreos/etcd/auth/authpb"
 	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
 	"github.com/coreos/etcd/mvcc/backend"
+
+	"github.com/stretchr/testify/assert"
```
Comment (xiang90) [inline auth/store_test.go]: let's not introduce this pkg please.

Comment (Rushit) [inline auth/store_test.go]: ok. I will refactor once #7234 to get merge.

**Comment by Rushit** [comment]
Updated PR with suggested changes by @mitake & @xiang90

File: auth/store_test.go [hunk 9]
```diff
@@ -213,3 +217,191 @@ func TestUserGrant(t *testing.T) {
...
+		t.Fatalf("expected user object, got %v", err)
+	}
+	expected := []string{"role-test"}
+	if !reflect.DeepEqual(expected, u.Roles) {
+		t.Fatalf("expected %v, got %v", expected, u.Roles)
+	}
+}
+
+func contains(array []string, str string) bool {
+	set := make(map[string]struct{}, len(array))
```
Comment (mitake) [inline auth/store_test.go]: Do we need to create the map? How about just iterating `array` and compare its element with `str`?

Comment (Rushit) [inline auth/store_test.go]: done.

File: auth/store_test.go [hunk 10]
```diff
@@ -213,3 +218,190 @@ func TestUserGrant(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	u, err := as.UserGet(&pb.AuthUserGetRequest{Name: "foo"})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if u == nil {
+		t.Fatalf("expected user object, got %v", err)
```
Comment (xiang90) [inline auth/store_test.go]: we already print err at line 232.

t.Fatal("expect user not nil, got nil")

Comment (xiang90) [inline auth/store_test.go]: can we also compare user name?

File: auth/store_test.go [hunk 11]
```diff
@@ -213,3 +218,190 @@ func TestUserGrant(t *testing.T) {
...
+	}
+	expected := []string{"role-test"}
+	if !reflect.DeepEqual(expected, u.Roles) {
+		t.Fatalf("expected %v, got %v", expected, u.Roles)
+	}
+}
+
+func contains(array []string, str string) bool {
```
Comment (xiang90) [inline auth/store_test.go]: can you move this helper function to be bottom of the test file?

File: auth/store_test.go [hunk 12]
```diff
@@ -213,3 +218,190 @@ func TestUserGrant(t *testing.T) {
...
+	for _, s := range array {
+		if strings.Compare(s, str) == 0 {
```
Comment (xiang90) [inline auth/store_test.go]: use s == str.

File: auth/store_test.go [hunk 13]
```diff
@@ -213,3 +218,190 @@ func TestUserGrant(t *testing.T) {
...
+		PermType: authpb.WRITE,
+		Key:      []byte("Keys"),
+		RangeEnd: []byte("RangeEnd"),
+	}
+	_, err = as.RoleGrantPermission(&pb.AuthRoleGrantPermissionRequest{
+		Name: "role-test-1",
+		Perm: perm,
+	})
+	if err != nil {
+		t.Error(err)
```
Comment (xiang90) [inline auth/store_test.go]: t.fatal

File: auth/store_test.go [hunk 14]
```diff
@@ -213,3 +218,190 @@ func TestUserGrant(t *testing.T) {
...
+	}
+
+	r, err := as.RoleGet(&pb.AuthRoleGetRequest{Role: "role-test-1"})
+	if err != nil {
+		t.Error(err)
```
Comment (xiang90) [inline auth/store_test.go]: fatal

File: auth/store_test.go [hunk 15]
```diff
@@ -213,3 +218,190 @@ func TestUserGrant(t *testing.T) {
...
+	}
+
+	r, err = as.RoleGet(&pb.AuthRoleGetRequest{Role: "role-test-1"})
+	if err != nil {
+		t.Error(err)
+	}
+	if len(r.Perm) != 0 {
+		t.Fatalf("expected %v, got %v", 0, len(r.Perm))
```
Comment (xiang90) [inline auth/store_test.go]: t.error

basically in the test setup phase, we use fatal. in the test case, we use error.

File: auth/store_test.go (hunk 10 above)
Comment (Rushit) [inline auth/store_test.go]: AuthUserGetResponse does not have user name.

File: auth/store_test.go [hunk 16]
```diff
@@ -213,3 +218,190 @@ func TestUserGrant(t *testing.T) {
...
+	r, err := as.RoleGet(&pb.AuthRoleGetRequest{Role: "role-test-1"})
+	if err != nil {
+		t.Error(err)
+	}
+
+	if !reflect.DeepEqual(perm, r.Perm[0]) {
+		t.Fatalf("expected %v, got %v", perm, r.Perm[0])
+	}
+
+	_, err = as.RoleRevokePermission(&pb.AuthRoleRevokePermissionRequest{
```
Comment (xiang90) [inline auth/store_test.go]: why do we test revoke in the grant test? i feel this is not necessary.

File: auth/store_test.go (hunk 15 above)
Comment (Rushit) [inline auth/store_test.go]: got it.

File: auth/store_test.go (hunk 16 above)
Comment (xiang90) [inline auth/store_test.go]: oh. sorry. i misread the test name.

File: auth/store_test.go [hunk 17]
```diff
@@ -213,3 +218,190 @@ func TestUserGrant(t *testing.T) {
...
+	_, err = as.UserRevokeRole(&pb.AuthUserRevokeRoleRequest{Name: "foo", Role: "role-test"})
+	if err != nil {
+		t.Fatal(err)
+	}
+	u, err := as.UserGet(&pb.AuthUserGetRequest{Name: "foo"})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if len(u.Roles) != 0 {
+		t.Fatalf("expected %v, got %v", 0, len(u.Roles))
```
Comment (xiang90) [inline auth/store_test.go]: t.errorf.

File: auth/store_test.go [hunk 18]
```diff
@@ -213,3 +218,190 @@ func TestUserGrant(t *testing.T) {
...
+	if len(r.Perm) != 0 {
+		t.Fatalf("expected %v, got %v", 0, len(r.Perm))
+	}
+}
+
+func TestRoleRevokePermission(t *testing.T) {
+	as, tearDown := setupAuthStore(t)
+	defer tearDown(t)
+
+	_, err := as.UserGrantRole(&pb.AuthUserGrantRoleRequest{User: "foo", Role: "role-test"})
```
Comment (xiang90) [inline auth/store_test.go]: can we grant two roles? revoke one? see if only one exists?

File: auth/store_test.go [hunk 19]
```diff
@@ -213,3 +218,190 @@ func TestUserGrant(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	rl, err := as.RoleList(&pb.AuthRoleListRequest{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	expected := []string{"root"}
+	if !reflect.DeepEqual(expected, rl.Roles) {
+		t.Fatalf("expected %v, got %v", expected, rl.Roles)
```
Comment (xiang90) [inline auth/store_test.go]: expected role = , got

File: auth/store_test.go (hunk 16 above)
Comment (Rushit) [inline auth/store_test.go]: it test both grant and revoke. I could separate out i guess.

File: auth/store_test.go [hunk 20]
```diff
@@ -213,3 +218,190 @@ func TestUserGrant(t *testing.T) {
...
+		t.Fatalf("expected %v, got %v", ErrInvalidAuthToken, err)
+	}
+
+	ctx = metadata.NewContext(context.Background(), metadata.New(map[string]string{"token": resp.Token}))
+	ai, err = as.AuthInfoFromCtx(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if ai.Username != "foo" {
+		t.Fatalf("expected %v, got %v", "foo", ai.Username)
```
Comment (xiang90) [inline auth/store_test.go]: use errorf in the test case.

File: auth/store_test.go (hunk 16 above)
Comment (xiang90) [inline auth/store_test.go]: if would be great if we can separate them out.

**Comment by xiang90** [comment]
@Rushit Thanks a lot for helping us improve the test coverage. This is super important for project health.

The test cases look good in general. I will let @mitake take the final look, and help you get it merged.

**Comment by Rushit** [comment]
I have latest push with suggested changes. I am hoping to get bette with each PR.  
No problem, Glad to contribute here 😄

File: auth/store_test.go (hunk 14 above)
Comment (Rushit) [inline auth/store_test.go]: updated.

**Comment by mitake** [comment]
@Rushit lgtm, thanks!
